src/main/java/uk/ac/ebi/ampt2d/accession/service/MonotonicAccessionGenerator.java

```diff
@@ -0,0 +1,241 @@
+            int remainingAccessionsToGenerate = totalAccessionsToGenerate - i;
+            MonotonicRange monotonicRange = pollNextMonotonicRange(remainingAccessionsToGenerate);
+            accessionRanges.add(monotonicRange);
+            i += monotonicRange.getTotalOfValues();
+        }
+        return accessionRanges;
+    }
+
+    /**
+     * Polls the next monotonic range to use, if the next available range contains more values than needed, splits
+     * the range and returns the excess to available ranges.
+     *
+     * @param totalValuesNeeded
+     * @return
+     */
+    private synchronized MonotonicRange pollNextMonotonicRange(int totalValuesNeeded) {
+        MonotonicRange monotonicRange = availableRanges.poll();
+        if (monotonicRange.getTotalOfValues() > totalValuesNeeded) {
+            Pair<MonotonicRange, MonotonicRange> splitResult = monotonicRange.split(totalValuesNeeded);
+            monotonicRange = splitResult.getFirst();
+            availableRanges.add(splitResult.getSecond());
+        }
+        return monotonicRange;
+    }
+
+    /**
+     * Ensures that the available ranges queue hold @param totalAccessionsToGenerate or more elements
+     *
+     * @param totalAccessionsToGenerate
+     */
+    private synchronized void increaseAvailableRangesUntilSizeIs(int totalAccessionsToGenerate) {
+        while (availableRanges.getTotalOfValuesInQueue() < totalAccessionsToGenerate) {
+            try {
+                ExponentialBackOff.execute(() -> reserveNewBlock(categoryId, instanceId, blockSize), 10, 30);
+            } catch (RuntimeException e) {
+                // Ignore, max backoff have been reached, we will try again until we can reserve blocks
+            }
+        }
+    }
+
+    /**
+     * Initialize the block service loading previously reserved blocks
+     */
+    @Override
+    public void afterPropertiesSet() {
+        List<ContiguousIdBlock> uncompletedBlocks = getUncompletedBlocksForThisInstanceOrdered();
+        //Insert as available ranges
+        for (ContiguousIdBlock block : uncompletedBlocks) {
+            addBlock(block);
+        }
+    }
+
+    /**
+     * TODO this can be changed to a query to recover and avoid the filter
+     */
+    private List<ContiguousIdBlock> getUncompletedBlocksForThisInstanceOrdered() {
+        try (Stream<ContiguousIdBlock> reservedBlocksOfThisInstance = contiguousIdBlockRepository
+                .findAllByCategoryIdAndInstanceIdOrderByEndAsc(categoryId, instanceId)) {
+            return reservedBlocksOfThisInstance.filter(ContiguousIdBlock::isNotFull).collect(Collectors.toList());
+        }
+    }
+
+    @Transactional(isolation = Isolation.SERIALIZABLE)
+    protected synchronized void reserveNewBlock(String categoryId, String instanceId, long size) {
+        ContiguousIdBlock lastBlock = contiguousIdBlockRepository.findFirstByCategoryIdOrderByEndDesc(categoryId);
+        if (lastBlock == null) {
```

Comment:
Change order of conditions because the positive branch will be executed only once per category and instance.